dev/core#2817 Add CaseToken processor

eileenmcnaughton opened the pull request:
This should be step 4 in https://lab.civicrm.org/dev/core/-/issues/2817 - it incorporates 1-3 which are also in separate PRs & should be mergeable 

This shows a discrepancy on date field handling however

- on the left the dates are formatted for the date fields (end_date, start_date) but not for created_date, modified_date. On the right they are formatted for both.  I'm inclined to think the left is an oversight & these are kinda obscure tokens but maybe we need an extension for those who Do want raw? We could move the other legacy tokens in there too from `replaceCaseTokens` - pondering

- Also on the left we have 'No' for the Boolean & FALSE on the right. I'm thinking that on the right is an oversight (since it is newer code) and we should fix the token processor to render a string for bool

UPDATE - I think we should stop providing 'is_deleted' as a token - that would punt on the boolean question & I don't think it makes sense (it is there by oversight IMHO)

![image](https://user-images.githubusercontent.com/336308/132194921-2b8ae3b2-5764-465a-ac28-21d52d67f17f.png)

It's also worth noting the function is only reachable from the send email & pdf search actions

mattwire commented:
For dates maybe we can come up with a generic solution that works on all date fields? Eg. `case.start_date:raw`

In this case I'm sure it's just an oversight and all 4 dates should be formatted according to the configured method.
Also the boolean should be formatted but that could also support `:raw` to output 0 / 1?

eileenmcnaughton commented:
@mattwire (warning - long thinking through things ramble follows)

So I think ideally we would have started with the values being 'raw' in their raw token form but formatting would be addable. This has been a problem with the pseudoconstants & the work around has been to start to  add in `{contribution.contribution_status_id:label}` to disambiguate and I think we can do similar with the booleans (for average users we have the advantage that we are advertising the tokens through a widget & fortunately we don't have *that many* fields to deal with.

Dates are trickier though IMHO. 

a) we need more flexibility with them & 
b) the raw namespace has been thoroughly 'taken'

The issue I logged is https://lab.civicrm.org/dev/core/-/issues/2746

As of last night I was leaning towards the option of supporting yet another format in the token processor (which is not hard technically but @totten is a bit alergic to the idea of the processor becoming too much like a templating language). It would look like

`{contribution.receive_date|}` = raw  
`{contribution.receive_date|date_format:'%d %b %Y %H:%M:%S'}` = formatted with the given string (same as smarty does)
`{contribution.receive_date|crm_date_format:'short'}` = formatted with a civi short date format


- adding the | for 'raw' would indicate the new syntax & we would switch to 'advertising' (in the widget) `{contribution.receive_date|crm_date_format:'short'}`  by default

The main argument against this approach is @totten's "we are writing our own templating language" one. Thinking down that path we had been wondering about the 'lazy load' option. The idea here is a new type of apiv4 result object which would be a lazy load object (Tim has some working? code in a gist somewhere).

In my imaginging (not sure where Tim's work is) it  would be something like

```
  $lazyLoadObject = Contact::get()->setSelect('id')->....->execute()->getLazyLoadObject();
  // Note we can access fields that were not originally selected
  $displayName = $lazyLoadObject['display_name']);
```

The thinking was then we could assign this object to smarty & smarty could do the formatting work e.g

`{$contact.birth_date|date_format:'%d %b %Y %H:%M:%S}`

I go back and forth on this. When I look at what I suggested above in terms of implementing as tokens it seems really simple to implement ( a couple of hours of work with tests) and it's really compatibile with our existing token expectations system.

I think @totten is seeing us go down that slippery slope & saying 'hey maybe this is the line we should reconsider our approach at'. However, to switch to using

`{$contact.birth_date|date_format:'%d %b %Y %H:%M:%S}`

what would it look like

- at the user level it would be somewhat opaque in that we would simply change what is 'advertised' in the widgets
- at the usage level it require some bigger changes in terms of smarty being disabled currently for most sites in pdf letter and email tasks in order to avoid security escalation and we would have to address that first.
- I feel like there is a bit of a risk of 'other gotchas' in switching over but not too sure what they are.

Neither way forwards has a lot of compatibility breakage in that both would respect existing tokens - although we still have to resolve what to do where the tokens are inconsistent. I'm now leaning towards an upgrade message on these tokens & maybe also on the 'loss' of 'hidden case tokens' (in case people have been using tokens that work by virtue of apiv3 but were never advertised in the widget which feels like a reasonable definition of 'supported'

mattwire commented:
@eileenmcnaughton Generally I don't think raw tokens are much use - the idea of the tokens is that they display something useful to the end-user and the raw values are not usually useful unless you're a developer.
Of course the difficulty is decided *how* things should be formatted by default and if they should be customisable. I do see a real benefit in allowing multiple date formats because you might choose to display time only or date only / day of week etc. eg. "Event is this Wednesday".

Tokens provide a simple way to customise messages *without* relying on a template language and I feel like expanding that to support date formatting would probably be a good idea as long as it has clear limits - anything complex *could* be handled via a raw token and the template language (assuming tokens are rendered before the template language is run). This provides passive support for other template languages such as twig/MJML because we're not adding any dependency on smarty for tokens.

eileenmcnaughton commented:
@mattwire 

I think you rightly point out there are 2 audiences for tokens - end users and those writing templates with logic - e.g workflow templates - e.g you allude to that with "anything complex could be handled via a raw token and the template language (assuming tokens are rendered before the template language is run)."

I just added a comment to https://lab.civicrm.org/dev/core/-/issues/2788 about a conversation we had that wasn't documented previously - but where we talked about being able to filter the tokens presented by audience

I'm thinking to put a note about this in the dev digest & maybe mid next week have agreed a format & we can implement. (As is usually the case the code to implement is the easy part)

eileenmcnaughton commented:
I just rebase this & #21327 (which is contribution recur) - there are fixes to the shared code to make this work & also to make contribution recur work. I consolidated the fixes (apart from the one that is specific to case) into that PR so that if it gets merged first all the fixes are merged & this becomes very simple

seamuslee001 commented:
Coleman also just noted that if you were creating a url using a token that linked to say a search kit search or something that accepted url values you would want to have the raw value.

eileenmcnaughton commented:
So here is where I think this is at - we have 3 issues

1) what should happen for date formatting generally. I think this PR helped us have that conversation & @totten is working up an idea about using the smarty modifiers within the token processor class.

2) what should happen with formatting specifically for created_date,  modified date and is_deleted in this PR. At this stage I'm inclined to do the following 
   a)  fix tests to pass with the change that this PR is making to those 
   b) merge an explainer update to docs https://lab.civicrm.org/documentation/docs/sysadmin/-/merge_requests/318 
   c) add an upgrade message to sites with civicase enabled pointing them to it

3) I've been worried about @demeritcowboy's comments that `replaceCaseTokens` would also have been replacing `hidden` tokens. However, I just checked & decide it is a non-issue as the 2 ;hidden tokens' (`contact_id` and `client_id`) resolve to arrays which are not usable anyway. So I think that simplifies my plan to switch over in gitlab.

![image](https://user-images.githubusercontent.com/336308/132781253-0e694d7f-1383-4461-939d-0d1997375f7b.png)

@demeritcowboy @seamuslee001 @mattwire are you on board with my plan in 2?

demeritcowboy commented:
Thanks for the deep dive. My comments:

* is_deleted I have no opinion. Never seen it used in a template.
* Date format for created/modified - I've also never seen those used in a template.
* Upgrade message - sure.
* client_id / contact_id - I'd say see https://github.com/civicrm/org.civicrm.casetokens which gives you tokens for a bunch of stuff about the client, which except for maybe {case.id} I'd argue is more useful than any of the core case tokens but that's another topic. My original comment about that was noting a discrepancy in the dropdown list, but it's available another way.

eileenmcnaughton commented:
@demeritcowboy `{case.id}` is a core token - see https://github.com/civicrm/civicrm-core/blob/2f5c0408165e4d80e3aadfc6b3b374a0f106840e/tests/phpunit/CRM/Utils/TokenConsistencyTest.php#L116

Thanks for the link to casetokens - I could link to it in the doc update MR but I suspect it's enough as is & if I just update the tests & add the upgrade message this should be mergeable

eileenmcnaughton commented:
I've updated #21327 to fix boolean handling such that they work like 'other fields where api v4 handles labels' - ie it ensures that 

`{case.is_deleted.label}` works

and I've updated this (which includes #21327) with the other changes discussed

eileenmcnaughton commented:
This has a green tick on the listing but not here - will re-run tests to see if it gets one

eileenmcnaughton commented:
test this please

demeritcowboy commented on CRM/Utils/Token.php:
```diff
@@ -1829,7 +1829,7 @@ public static function getCustomFieldTokens($entity, $usedForTokenWidget = FALSE
     $tokenName = $usedForTokenWidget ? "{contribution.custom_%d}" : "custom_%d";
     foreach (CRM_Core_BAO_CustomField::getFields($entity) as $id => $info) {
-      $customTokens[sprintf($tokenName, $id)] = $info['label'];
+      $customTokens[sprintf($tokenName, $id)] = $info['label'] . ' :: ' . $info['groupTitle'];
```
The other recent PRs about custom field choices were putting grouptitle first.

eileenmcnaughton replied:
@demeritcowboy - it looks like 'second' is consistent with other token places - eg. https://github.com/civicrm/civicrm-core/pull/21429/files#diff-c3c60600f080b49f2a9029b82865098c2a2742fbcedce0ecd731376917d80e1bR495 reflects 'current behaviour'

I'm happy to look at flipping but I think I'd rather do it in a follow up if possible because I think there might be a couple of moving parts here

demeritcowboy commented:
Putting merge-ready since running seems to be ok.
I'm a little confused about the boolean handling since the ones being removed seem to appear anyway (like is_test), nothing gets removed for case tokens, and then when you submit there are some extra ones removed that aren't in the list to begin with before or after the patch (like auto_renew). But it sounds like there's still some ongoing discussion in this area.

eileenmcnaughton commented:
@demeritcowboy yeah the boolean is a bit of a side issue here - I needed some sort of resolution to get the test to pass - but the field affected is 'is_deleted' and also some fields that weren't previously exposed.

The goal is to advertise the label is '{case.is_deleted:label}' in the widget as that is unambiguous

eileenmcnaughton commented:
Also - if I can get a merge on this or https://github.com/civicrm/civicrm-core/pull/21429 today then I will do the next steps on those today -(for this one it is more moving to the pdf trait - for the other it's fixing handling)  otherwise I will try to pick up again next window

colemanw commented:
This has had "merge-ready" for a few days and includes unit test coverage. Time to merge.

demeritcowboy commented:
It was only merge-ready for a few hours on a Sunday, but I still haven't figured timezones out and how it's already tomorrow today, and the wasps here are buzzing about like crazy this year so it wouldn't surprise me if it's creating a temporal distortion of some kind.

Seriously, what I was confused about was how it doesn't seem to make the same list when generating the dropdown as on submit, and for Case it doesn't even hit that Boolean code when generating the dropdown. But it works, and there's obviously more coming.

eileenmcnaughton commented:
@demeritcowboy we need to consolidate the function to list tokens - I added a separate PR for that - but I'm gonna do the PRs to switch pdf & email over before really digging in

colemanw commented:
You're right @demeritcowboy I misread the time the merge-ready label was added. But I trust Eileen to do any necessary followup if there are issues.

eileenmcnaughton commented:
It's Monday here :-)

eileenmcnaughton commented:
@demeritcowboy ah now I know why I left https://github.com/civicrm/civicrm-core/pull/21446 deliberately out of scope. 

This PR is about the tokens that can be rendered by the processor. However, we want to transitionally hide the ambiguous ones eg `{case.case_type_id}` and promote the unambiguous alternative `{case.case_type_id:label}`. This leaves us with the widget not returning the full set - and not returning the 'preferred' `{case.is_deleted:label}`. If it had been a different field I would have added the preferred version to the widget. I still could edit it manually but I guess I mostly just feel like it's a field that probably has no use as a token - and we should focus on the bigger picture for 'advertising' - https://lab.civicrm.org/dev/core/-/issues/2788 - ie. how do we somewhat limit tokens to a useful set.
